### PR TITLE
STITCH-1950 Add collection level sync configuration

### PR DESCRIPTION
**Pull request by jsflax**
*Notes: I've handled the possible null case here in what I felt was best for usability, however, it may need to be agreed upon.* 

Description:
Configuration for sync happens at collection.sync().configure() where configure is: configure(<ConflictHandler>, <ChangeEventListener>, <ErrorListener>). See spec for more info.

This configure should be the signal to the data synchronizer to start synchronzing if it hasn't already. Subsequent configurations just update handlers and listeners. The signal is to prevent an error condition where we were to sync but do not know how to handle conflcits since a handler has not been set up yet. Most app code will always be calling configure first.

### CODE REVIEW

**Comment by jsflax** [comment]
Seems like some formatting got included in this PR. I can remove it, or we can keep it if we want.

**Comment by jsflax** [comment]
Evg is being flakey. Otherwise this is ready for review. The driveby tab changes are valid, however, I can move them to another PR if we feel that it's here inappropriately.

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java [hunk 1]
```diff
@@ -195,24 +199,47 @@ public void reloadConfig() {
...
     this.errorListener = errorListener;
+
     this.syncConfig.getNamespaceConfig(namespace).configure(
         conflictHandler,
         changeEventListener,
         codec
     );
-    this.triggerListeningToNamespace(namespace);
+
+    if (!this.isConfigured) {
```
Comment (edaniels) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java]: [minor] We should grab the lock first before checking configured otherwise this can be a torn read or write.

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java [hunk 2]
```diff
@@ -195,24 +199,47 @@ public void reloadConfig() {
...
+      syncLock.lock();
+      try {
+        this.isConfigured = true;
+      } finally {
+        syncLock.unlock();
+      }
+      this.triggerListeningToNamespace(namespace);
+    }
+
+    if (!isRunning) {
```
Comment (edaniels) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java]: [q] Why is this piece of code here? Seems irrelevant

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java [hunk 3]
```diff
@@ -1006,7 +1043,7 @@ private void resolveConflict(
     final boolean acceptRemote =
         (remoteEvent.getFullDocument() == null && resolvedDocument == null)
             || (remoteEvent.getFullDocument() != null
-            && transformedRemoteEvent.getFullDocument().equals(resolvedDocument));
+            && remoteEvent.getFullDocument().equals(resolvedDocument));
```
Comment (edaniels) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java]: [major] This is reintroducing a bug. Please undo any reformats or changes unrelated to this ticket in this PR.

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java [hunk 4]
```diff
@@ -103,6 +105,8 @@
 
   private InstanceSynchronizationConfig syncConfig;
   private boolean syncThreadEnabled = true;
+  private boolean isConfigured = false;
+  private boolean isRunning = false;
```
Comment (edaniels) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java]: Should be able to get rid of this.

File: server/services/mongodb-remote/src/test/java/com/mongodb/stitch/server/services/mongodb/remote/internal/SyncMongoClientIntTests.kt [hunk 5]
```diff
@@ -899,6 +902,27 @@ class SyncMongoClientIntTests : BaseStitchServerIntTest() {
...
+    fun testConfigure() {
+        val testSync = getTestSync()
+
+        (mongoClient as RemoteMongoClientImpl).dataSynchronizer.enableSyncThread()
+        assertFalse((mongoClient as RemoteMongoClientImpl).dataSynchronizer.isRunning)
+
+        val docToInsert = Document("hello", "world")
+        testSync.insertOneAndSync(docToInsert)
+
+        assertFalse((mongoClient as RemoteMongoClientImpl).dataSynchronizer.isRunning)
```
Comment (edaniels) [inline server/services/mongodb-remote/src/test/java/com/mongodb/stitch/server/services/mongodb/remote/internal/SyncMongoClientIntTests.kt]: [test] Instead of needing an isRunning as var and exposed, you can instead just add a semaphore to watch a change stream come in.

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java (hunk 2 above)
Comment (edaniels) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java]: [nit] Should be able to get rid of this

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java [hunk 6]
```diff
@@ -1493,6 +1540,24 @@ private void triggerListeningToNamespace(final MongoNamespace namespace) {
     }
   }
 
+  /**
+   * Whether or not the DataSynchronizer is running in the background.
+   *
+   * @return true if running, false if not
+   */
+  public boolean isRunning() {
```
Comment (edaniels) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java]: [nit] Should be able to get rid of this

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java [hunk 7]
```diff
@@ -1493,6 +1540,24 @@ private void triggerListeningToNamespace(final MongoNamespace namespace) {
...
+    return isRunning;
+  }
+
+  /**
+   * Whether or not the DataSynchronizer has been configured
+   *
+   * @return true if configured, false if not
+   */
+  boolean isConfigured() {
```
Comment (edaniels) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java]: [minor] must be `synchronized`

File: server/services/mongodb-remote/src/test/java/com/mongodb/stitch/server/services/mongodb/remote/internal/SyncMongoClientIntTests.kt [hunk 8]
```diff
@@ -899,6 +902,27 @@ class SyncMongoClientIntTests : BaseStitchServerIntTest() {
         }
     }
 
+    @Test
+    fun testConfigure() {
+        val testSync = getTestSync()
+
+        (mongoClient as RemoteMongoClientImpl).dataSynchronizer.enableSyncThread()
```
Comment (edaniels) [inline server/services/mongodb-remote/src/test/java/com/mongodb/stitch/server/services/mongodb/remote/internal/SyncMongoClientIntTests.kt]: I'm not quite positive, but it seems like the expose and implementation of `.enableSyncThread()` is not strictly necessary. You can use manual sync passes if you need to.

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java [hunk 9]
```diff
@@ -231,12 +258,22 @@ public void start() {
...
       if (syncThreadEnabled) {
         syncThread.start();
+        isRunning = true;
       }
     } finally {
       syncLock.unlock();
     }
   }
 
+  public void enableSyncThread() {
```
Comment (edaniels) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java]: [nit] Think this can be deleted.

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java [hunk 10]
```diff
@@ -195,24 +199,47 @@ public void reloadConfig() {
...
+      try {
+        this.isConfigured = true;
+      } finally {
+        syncLock.unlock();
+      }
+      this.triggerListeningToNamespace(namespace);
+    }
+
+    if (!isRunning) {
+      this.start();
```
Comment (edaniels) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java]: I think this is causing your flakiness now. You have a BG thread running. Removing this will fix that.

File: core/services/mongodb-remote/src/test/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreDocumentSynchronizationConfigUnitTests.kt [hunk 11]
```diff
@@ -38,41 +38,41 @@ class CoreDocumentSynchronizationConfigUnitTests {
...
-                    System.getProperty("java.io.tmpdir"),
-                    "<local-app-name>",
-                    "<local-app-version>",
-                    BsonUtils.DEFAULT_CODEC_REGISTRY,
-                    networkMonitor,
-                    authMonitor
-            ),
-            "mongodblocal",
-            ServerEmbeddedMongoClientFactory.getInstance()
+        StitchAppClientInfo(
```
Comment (edaniels) [inline core/services/mongodb-remote/src/test/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreDocumentSynchronizationConfigUnitTests.kt]: Looks like this file can be reverted.

File: core/services/mongodb-remote/src/test/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizerUnitTests.kt [hunk 12]
```diff
@@ -121,4 +135,68 @@ class DataSynchronizerUnitTests {
...
+        assertFalse(dataSynchronizer.isRunning)
+
+        // mock the necessary config args
+        val conflictHandler = mock(ConflictHandler::class.java) as ConflictHandler<BsonDocument>
+        val changeEventListener = mock(ChangeEventListener::class.java) as ChangeEventListener<BsonDocument>
+        val errorListener = mock(ErrorListener::class.java)
+        val bsonCodec = BsonDocumentCodec()
+
+        // fetch a new namespace, creating a new config
+        val nsConfig: NamespaceSynchronizationConfig = dataSynchronizer.getNamespaceConfig(namespace)
```
Comment (edaniels) [inline core/services/mongodb-remote/src/test/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizerUnitTests.kt]: [nit] I think you can get rid of `dataSynchronizer.getNamespaceConfig` and `getNamespaceConfig` as being exposed. Instead you should be able to utilize the top level public methods on the data synchronizer instead to act as a normal consumer which would make this a stronger test.

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java (hunk 4 above)
Comment (jsflax) [reply core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java]: Spoken about offline, resolving

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java (hunk 1 above)
Comment (jsflax) [reply core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java]: 👍

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java (hunk 2 above)
Comment (jsflax) [reply core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java]: Resolved offline.

Comment (jsflax) [reply core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java]: Resolved offline.

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java (hunk 9 above)
Comment (jsflax) [reply core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java]: Having this here allows us to check that the thread is started from within our tests.

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java (hunk 6 above)
Comment (jsflax) [reply core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java]: Resolved offline.

File: server/services/mongodb-remote/src/test/java/com/mongodb/stitch/server/services/mongodb/remote/internal/SyncMongoClientIntTests.kt (hunk 8 above)
Comment (jsflax) [reply server/services/mongodb-remote/src/test/java/com/mongodb/stitch/server/services/mongodb/remote/internal/SyncMongoClientIntTests.kt]: Alright. This makes more sense. It will slightly change the tests, but yes, if we address your comment below, we can remove the `enableSyncThread` method.

File: core/services/mongodb-remote/src/test/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizerUnitTests.kt (hunk 12 above)
Comment (jsflax) [reply core/services/mongodb-remote/src/test/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizerUnitTests.kt]: Might it make more sense to do that in an integration test?

Comment (edaniels) [reply core/services/mongodb-remote/src/test/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizerUnitTests.kt]: I don't think so. The unit test here is aiming to check what happens if you are to configure twice. The first configuration should end up starting the listener. We don't need to reach that far into the implementation details. It ends up making the test brittle and not covering the actual aspect you want to test via the public API.

File: server/services/mongodb-remote/src/test/java/com/mongodb/stitch/server/services/mongodb/remote/internal/SyncMongoClientIntTests.kt (hunk 5 above)
Comment (jsflax) [reply server/services/mongodb-remote/src/test/java/com/mongodb/stitch/server/services/mongodb/remote/internal/SyncMongoClientIntTests.kt]: 👍

File: android/services/mongodb-remote/src/main/java/com/mongodb/stitch/android/services/mongodb/remote/Sync.java [hunk 13]
```diff
@@ -43,7 +45,7 @@
    *                         document.
    * @param errorListener the error listener to invoke when an irrecoverable error occurs
    */
-  void configure(final ConflictHandler<DocumentT> conflictHandler,
+  void configure(@NonNull final ConflictHandler<DocumentT> conflictHandler,
                  final ChangeEventListener<DocumentT> changeEventListener,
```
Comment (adamchel) [inline android/services/mongodb-remote/src/main/java/com/mongodb/stitch/android/services/mongodb/remote/Sync.java]: [nit] if we're labeling the conflict handler `@NonNull`, I guess we should explicitly mark the rest as `@Nullable`?

File: android/services/mongodb-remote/src/main/java/com/mongodb/stitch/android/services/mongodb/remote/internal/SyncImpl.java [hunk 14]
```diff
@@ -45,7 +47,7 @@
   }
 
   @Override
-  public void configure(final ConflictHandler<DocumentT> conflictHandler,
+  public void configure(@NonNull final ConflictHandler<DocumentT> conflictHandler,
```
Comment (adamchel) [inline android/services/mongodb-remote/src/main/java/com/mongodb/stitch/android/services/mongodb/remote/internal/SyncImpl.java]: [nit] ditto

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/CoreSync.java [hunk 15]
```diff
@@ -33,13 +35,13 @@
...
-   * @param conflictResolver the conflict resolver to invoke when a conflict happens between local
-   *                         and remote events.
+   * @param conflictHandler the conflict resolver to invoke when a conflict happens between local
+   *                        and remote events.
    * @param changeEventListener the event listener to invoke when a change event happens for the
    *                         document.
    * @param errorListener the error listener to invoke when an irrecoverable error occurs
    */
-  void configure(final ConflictHandler<DocumentT> conflictResolver,
+  void configure(@Nonnull final ConflictHandler<DocumentT> conflictHandler,
```
Comment (adamchel) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/CoreSync.java]: [nit] ditto

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreSyncImpl.java [hunk 16]
```diff
@@ -54,12 +56,12 @@ public CoreSyncImpl(final MongoNamespace namespace,
   }
 
   @Override
-  public void configure(final ConflictHandler<DocumentT> conflictResolver,
+  public void configure(@Nonnull final ConflictHandler<DocumentT> conflictHandler,
```
Comment (adamchel) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreSyncImpl.java]: [nit] ditto

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java [hunk 17]
```diff
@@ -195,24 +199,47 @@ public void reloadConfig() {
...
     } finally {
       syncLock.unlock();
     }
   }
 
-
-  public <T> void configure(final MongoNamespace namespace,
-                            final ConflictHandler<T> conflictHandler,
+  public <T> void configure(@Nonnull final MongoNamespace namespace,
+                            @Nonnull final ConflictHandler<T> conflictHandler,
```
Comment (adamchel) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java]: [nit] ditto

File: server/services/mongodb-remote/src/main/java/com/mongodb/stitch/server/services/mongodb/remote/Sync.java [hunk 18]
```diff
@@ -43,7 +45,7 @@
    * @param errorListener the error listener to invoke when an irrecoverable error happens
    *                      during data synchronization
    */
-  void configure(final ConflictHandler<DocumentT> conflictResolver,
+  void configure(@Nonnull final ConflictHandler<DocumentT> conflictResolver,
```
Comment (adamchel) [inline server/services/mongodb-remote/src/main/java/com/mongodb/stitch/server/services/mongodb/remote/Sync.java]: [nit] ditto

File: server/services/mongodb-remote/src/main/java/com/mongodb/stitch/server/services/mongodb/remote/internal/SyncImpl.java [hunk 19]
```diff
@@ -42,7 +44,7 @@
   }
 
   @Override
-  public void configure(final ConflictHandler<DocumentT> conflictResolver,
+  public void configure(@Nonnull final ConflictHandler<DocumentT> conflictResolver,
```
Comment (adamchel) [inline server/services/mongodb-remote/src/main/java/com/mongodb/stitch/server/services/mongodb/remote/internal/SyncImpl.java]: [nit] ditto
